File: packages/server-sdk-go/client_test.go
package prompthash

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"testing"
	"time"
)

type recordedRequest struct {
	Method        string
	Path          string
	RawQuery      string
	Authorization string
	AcceptVersion string
	ContentType   string
	Idempotency   string
	Body          string
}

type testServer struct {
	mu       sync.Mutex
	statuses []int
	bodies   []string
	requests []recordedRequest
	index    int
}

func (s *testServer) handler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)

		s.mu.Lock()
		i := s.index
		if i >= len(s.statuses) {
			i = len(s.statuses) - 1
		}
		status := s.statuses[i]
		body := s.bodies[i]
		s.index++
		s.requests = append(s.requests, recordedRequest{
			Method:        r.Method,
			Path:          r.URL.Path,
			RawQuery:      r.URL.RawQuery,
			Authorization: r.Header.Get("Authorization"),
			AcceptVersion: r.Header.Get("Accept-Version"),
			ContentType:   r.Header.Get("Content-Type"),
			Idempotency:   r.Header.Get("Idempotency-Key"),
			Body:          string(raw),
		})
		s.mu.Unlock()

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		_, _ = w.Write([]byte(body))
	}
}

func (s *testServer) call(index int) recordedRequest {
	s.mu.Lock()
	defer s.mu.Unlock()
	if index >= len(s.requests) {
		return recordedRequest{}
	}
	return s.requests[index]
}

func (s *testServer) count() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.requests)
}

type harness struct {
	client *Client
	server *httptest.Server
	rec    *testServer
	delays *[]time.Duration
}

func newHarness(t *testing.T, cfg Config, statuses []int, bodies []string) *harness {
	t.Helper()

	rec := &testServer{statuses: statuses, bodies: bodies}
	server := httptest.NewServer(rec.handler())
	t.Cleanup(server.Close)

	var delays []time.Duration
	cfg.BaseURL = server.URL
	if cfg.Sleep == nil {
		cfg.Sleep = func(_ context.Context, delay time.Duration) error {
			delays = append(delays, delay)
			return nil
		}
	}
	client, err := NewClient(cfg)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return &harness{client: client, server: server, rec: rec, delays: &delays}
}

func TestNewClientRequiresAValidBaseURL(t *testing.T) {
	if _, err := NewClient(Config{}); err == nil {
		t.Error("empty BaseURL should fail")
	}
	if _, err := NewClient(Config{BaseURL: "ftp://example.test"}); err == nil {
		t.Error("non-http BaseURL should fail")
	}
	if _, err := NewClient(Config{BaseURL: "https://api.example.test/"}); err != nil {
		t.Errorf("trailing slash should be accepted: %v", err)
	}
}

func TestRequestSendsAuthVersioningAndIdempotencyHeaders(t *testing.T) {
	h := newHarness(t, Config{
		APIKey:     "pm_abc123_supersecret",
		APIVersion: "2025-01-01",
	}, []int{200}, []string{`{"ok":true}`})

	err := h.client.Post(context.Background(), "/api/webhooks", RequestOptions{
		Body:           map[string]string{"walletAddress": "GABC"},
		IdempotencyKey: "idem-1",
	}, nil)
	if err != nil {
		t.Fatalf("Post: %v", err)
	}

	call := h.rec.call(0)
	if call.Method != http.MethodPost {
		t.Errorf("method = %q", call.Method)
	}
	if call.Authorization != "Bearer pm_abc123_supersecret" {
		t.Errorf("authorization = %q", call.Authorization)
	}
	if call.AcceptVersion != "2025-01-01" {
		t.Errorf("accept-version = %q", call.AcceptVersion)
	}
	if call.ContentType != "application/json" {
		t.Errorf("content-type = %q", call.ContentType)
	}
	if call.Idempotency != "idem-1" {
		t.Errorf("idempotency-key = %q", call.Idempotency)
	}
	var payload map[string]string
	if err := json.Unmarshal([]byte(call.Body), &payload); err != nil {
		t.Fatalf("Unmarshal: %v", err)
	}
	if payload["walletAddress"] != "GABC" {
		t.Errorf("body = %q", call.Body)
	}
}

func TestOmitsAuthorizationWithoutAnAPIKey(t *testing.T) {
	h := newHarness(t, Config{}, []int{200}, []string{`{}`})

	if err := h.client.Get(context.Background(), "/api/prompts", RequestOptions{}, nil); err != nil {
		t.Fatalf("Get: %v", err)
	}
	if call := h.rec.call(0); call.Authorization != "" {
		t.Errorf("authorization = %q", call.Authorization)
	}
}

func TestListPromptsBuildsTheQueryString(t *testing.T) {
	h := newHarness(t, Config{}, []int{200}, []string{`{"total":0}`})

	if _, err := h.client.ListPrompts(context.Background(), ListPromptsParams{
		Page:  2,
		Limit: 10,
		Sort:  "upvotes",
	}); err != nil {
		t.Fatalf("ListPrompts: %v", err)
	}

	call := h.rec.call(0)
	if call.Path != "/api/prompts" {
		t.Errorf("path = %q", call.Path)
	}
	values, err := url.ParseQuery(call.RawQuery)
	if err != nil {
		t.Fatalf("ParseQuery: %v", err)
	}
	if values.Get("page") != "2" || values.Get("limit") != "10" || values.Get("sort") != "upvotes" {
		t.Errorf("query = %q", call.RawQuery)
	}
}

func TestTypedErrorCarriesTheMachineReadableCode(t *testing.T) {
	h := newHarness(t, Config{}, []int{403}, []string{
		`{"apiVersion":"2025-01-01","error":"Prompt access has not been purchased.","code":"ACCESS_NOT_PURCHASED"}`,
	})

	err := h.client.GetPrompt(context.Background(), "abc")
	if err == nil {
		t.Fatal("expected an error")
	}

	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.Status != 403 || apiErr.Code != CodeAccessNotPurchased {
		t.Errorf("error = %+v", apiErr)
	}
	if apiErr.Retryable() {
		t.Error("ACCESS_NOT_PURCHASED must not be retryable")
	}
	if !strings.Contains(apiErr.Message, "not been purchased") {
		t.Errorf("message = %q", apiErr.Message)
	}
	if h.rec.count() != 1 {
		t.Errorf("requests = %d, want 1", h.rec.count())
	}
}

func TestDoesNotRetryAValidationFailure(t *testing.T) {
	h := newHarness(t, Config{}, []int{400}, []string{
		`{"error":"Missing fields","code":"MISSING_FIELDS"}`,
	})

	err := h.client.Post(context.Background(), "/api/webhooks", RequestOptions{
		Body: map[string]string{},
	}, nil)
	if err == nil {
		t.Fatal("expected an error")
	}

	var apiErr *APIError
	if !errors.As(err, &apiErr) || apiErr.Code != CodeMissingFields {
		t.Fatalf("expected MISSING_FIELDS *APIError, got %T: %v", err, err)
	}
	if h.rec.count() != 1 {
		t.Errorf("requests = %d, want 1", h.rec.count())
	}
}

func TestRetries429UsingTheResetTimestamp(t *testing.T) {
	reset := time.Now().Add(5 * time.Second).UnixMilli()
	h := newHarness(t, Config{}, []int{429, 200}, []string{
		`{"error":"Too many requests.","code":"RATE_LIMIT_IP","reset":` + strconv.FormatInt(reset, 10) + `}`,
		`{"prompts":[]}`,
	})

	var page MarketplacePage
	if err := h.client.Get(context.Background(), "/api/prompts", RequestOptions{}, &page); err != nil {
		t.Fatalf("Get: %v", err)
	}

	if h.rec.count() != 2 {
		t.Errorf("requests = %d, want 2", h.rec.count())
	}
	if len(*h.delays) != 1 {
		t.Fatalf("delays = %d, want 1", len(*h.delays))
	}
	if delay := (*h.delays)[0]; delay < 4*time.Second || delay > 5*time.Second {
		t.Errorf("delay = %v, want ~5s", delay)
	}
}

func TestRetriesTransientServerErrorsWithBackoff(t *testing.T) {
	h := newHarness(t, Config{
		RetryBaseDelay: 100 * time.Millisecond,
		RetryMaxDelay:  time.Second,
	}, []int{503, 503, 200}, []string{
		`{"error":"degraded"}`,
		`{"error":"degraded"}`,
		`{"ok":true}`,
	})

	if err := h.client.Get(context.Background(), "/api/prompts", RequestOptions{}, nil); err != nil {
		t.Fatalf("Get: %v", err)
	}

	if h.rec.count() != 3 {
		t.Errorf("requests = %d, want 3", h.rec.count())
	}
	delays := *h.delays
	if len(delays) != 2 {
		t.Fatalf("delays = %d, want 2", len(delays))
	}
	if delays[0] < 100*time.Millisecond {
		t.Errorf("first delay = %v, want >= 100ms", delays[0])
	}
	if delays[1] < 200*time.Millisecond {
		t.Errorf("second delay = %v, want >= 200ms", delays[1])
	}
}

func TestRetryBudgetOfZeroDisablesRetries(t *testing.T) {
	h := newHarness(t, Config{}, []int{503}, []string{`{"error":"degraded"}`})
	zero := 0

	err := h.client.Get(context.Background(), "/api/prompts", RequestOptions{
		RetryBudget: &zero,
	}, nil)
	if err == nil {
		t.Fatal("expected an error")
	}
	if h.rec.count() != 1 {
		t.Errorf("requests = %d, want 1", h.rec.count())
	}
	if len(*h.delays) != 0 {
		t.Errorf("delays = %d, want 0", len(*h.delays))
	}
}

func TestStopsAfterTheConfiguredRetryBudget(t *testing.T) {
	h := newHarness(t, Config{MaxRetries: 3}, []int{500}, []string{`{"error":"boom"}`})

	if err := h.client.Get(context.Background(), "/api/prompts", RequestOptions{}, nil); err == nil {
		t.Fatal("expected an error")
	}
	if h.rec.count() != 4 {
		t.Errorf("requests = %d, want 4", h.rec.count())
	}
	if len(*h.delays) != 3 {
		t.Errorf("delays = %d, want 3", len(*h.delays))
	}
}

func TestWrapsTransportFailuresInANetworkError(t *testing.T) {
	// Nothing listens on this port, so the request fails before any response.
	var delays []time.Duration
	client, err := NewClient(Config{
		BaseURL: "http://127.0.0.1:1",
		Sleep: func(_ context.Context, delay time.Duration) error {
			delays = append(delays, delay)
			return nil
		},
	})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}

	failure := client.Get(context.Background(), "/api/prompts", RequestOptions{}, nil)
	if failure == nil {
		t.Fatal("expected an error")
	}

	var netErr *NetworkError
	if !errors.As(failure, &netErr) {
		t.Fatalf("expected *NetworkError, got %T: %v", failure, failure)
	}
	if netErr.Method != http.MethodGet {
		t.Errorf("method = %q", netErr.Method)
	}
	if !strings.Contains(netErr.URL, "/api/prompts") {
		t.Errorf("url = %q", netErr.URL)
	}
	if len(delays) != 2 {
		t.Errorf("delays = %d, want 2", len(delays))
	}
}

func TestRegisterWebhookSendsTheSubscriptionBody(t *testing.T) {
	h := newHarness(t, Config{}, []int{201}, []string{
		`{"id":"w1","secret":"s3cret"}`,
	})

	registration, err := h.client.RegisterWebhook(context.Background(), RegisterWebhookParams{
		WalletAddress: "GABC",
		URL:           "https://example.test/hook",
		Events:        []string{"PromptPurchased"},
	})
	if err != nil {
		t.Fatalf("RegisterWebhook: %v", err)
	}
	if registration.Secret != "s3cret" {
		t.Errorf("secret = %q", registration.Secret)
	}

	var body map[string]any
	if err := json.Unmarshal([]byte(h.rec.call(0).Body), &body); err != nil {
		t.Fatalf("Unmarshal: %v", err)
	}
	if body["walletAddress"] != "GABC" || body["url"] != "https://example.test/hook" {
		t.Errorf("body = %v", body)
	}
}

func TestListWebhookDeliveriesScopesByWallet(t *testing.T) {
	h := newHarness(t, Config{}, []int{200}, []string{`[]`})

	if _, err := h.client.ListWebhookDeliveries(context.Background(), "GABC"); err != nil {
		t.Fatalf("ListWebhookDeliveries: %v", err)
	}

	values, err := url.ParseQuery(h.rec.call(0).RawQuery)
	if err != nil {
		t.Fatalf("ParseQuery: %v", err)
	}
	if values.Get("walletAddress") != "GABC" {
		t.Errorf("query = %q", h.rec.call(0).RawQuery)
	}
}

func TestGetPromptEscapesPathSegments(t *testing.T) {
	h := newHarness(t, Config{}, []int{200}, []string{`{"id":"1"}`})

	if _, err := h.client.GetPrompt(context.Background(), "prompt 1/2"); err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}

	if call := h.rec.call(0); call.Path != "/api/prompts/prompt 1/2" {
		t.Errorf("decoded path = %q", call.Path)
	}
}

File: packages/server-sdk/test/client.test.ts
import { describe, expect, it } from "vitest";
import { PromptHashServerClient } from "../src/client.js";
import { PromptHashApiError, PromptHashNetworkError } from "../src/errors.js";
import { jsonResponse, stubSleep, stubTransport } from "./support.js";

function makeClient(
  transport: ReturnType<typeof stubTransport>,
  overrides: Record<string, unknown> = {},
) {
  const { sleep, delays } = stubSleep();
  const client = new PromptHashServerClient({
    baseUrl: "https://api.example.test/",
    apiKey: "pm_abc123_supersecret",
    apiVersion: "2025-01-01",
    fetch: transport.fetch,
    sleep,
    ...overrides,
  });
  return { client, delays };
}

describe("PromptHashServerClient request pipeline", () => {
  it("requires a baseUrl", () => {
    expect(() => new PromptHashServerClient({ baseUrl: "" })).toThrow(TypeError);
  });

  it("normalises the base URL and builds the query string", async () => {
    const transport = stubTransport(jsonResponse(200, { prompts: [], total: 0 }));
    const { client } = makeClient(transport);

    await client.listPrompts({
      page: 2,
      limit: 10,
      sort: "upvotes",
      search: undefined,
    });

    expect(transport.calls[0].url).toBe(
      "https://api.example.test/api/prompts?page=2&limit=10&sort=upvotes",
    );
  });

  it("omits null query values entirely", async () => {
    const transport = stubTransport(jsonResponse(200, {}));
    const { client } = makeClient(transport);

    await client.get("/api/prompts", {
      query: { cursor: null, page: 1 },
    });

    expect(transport.calls[0].url).toBe("https://api.example.test/api/prompts?page=1");
  });

  it("sends auth, versioning, and content headers", async () => {
    const transport = stubTransport(jsonResponse(200, {}));
    const { client } = makeClient(transport);

    await client.post("/api/webhooks", {
      body: { walletAddress: "GABC" },
      idempotencyKey: "idem-1",
    });

    const { init } = transport.calls[0];
    expect(init.method).toBe("POST");
    expect(init.headers.Authorization).toBe("Bearer pm_abc123_supersecret");
    expect(init.headers["Accept-Version"]).toBe("2025-01-01");
    expect(init.headers["Content-Type"]).toBe("application/json");
    expect(init.headers["Idempotency-Key"]).toBe("idem-1");
    expect(init.body).toBe(JSON.stringify({ walletAddress: "GABC" }));
  });

  it("lets per-call headers win over the defaults", async () => {
    const transport = stubTransport(jsonResponse(200, {}));
    const { client } = makeClient(transport, {
      defaultHeaders: { "X-Trace-Id": "abc", "Accept-Version": "2024-01-01" },
    });

    await client.get("/api/prompts", { headers: { "X-Trace-Id": "def" } });

    const { init } = transport.calls[0];
    expect(init.headers["X-Trace-Id"]).toBe("def");
    expect(init.headers["Accept-Version"]).toBe("2024-01-01");
  });

  it("returns raw text when the response is not JSON", async () => {
    const transport = stubTransport(jsonResponse(200, "pong"));
    const { client } = makeClient(transport);

    await expect(client.get("/health")).resolves.toBe("pong");
  });
});

describe("PromptHashServerClient error handling", () => {
  it("throws a typed error carrying the machine-readable code", async () => {
    const transport = stubTransport(
      jsonResponse(403, {
        apiVersion: "2025-01-01",
        error: "Prompt access has not been purchased.",
        code: "ACCESS_NOT_PURCHASED",
      }),
    );
    const { client, delays } = makeClient(transport);

    const failure = client.getPrompt("abc");

    await expect(failure).rejects.toBeInstanceOf(PromptHashApiError);
    await failure.catch((err: PromptHashApiError) => {
      expect(err.status).toBe(403);
      expect(err.code).toBe("ACCESS_NOT_PURCHASED");
      expect(err.retryable).toBe(false);
      expect(err.message).toContain("not been purchased");
    });
    expect(transport.calls).toHaveLength(1);
    expect(delays).toHaveLength(0);
  });

  it("does not retry a validation failure", async () => {
    const transport = stubTransport(
      jsonResponse(400, { error: "Missing fields", code: "MISSING_FIELDS" }),
    );
    const { client } = makeClient(transport);

    await expect(client.post("/api/webhooks", { body: {} })).rejects.toBeInstanceOf(
      PromptHashApiError,
    );
    expect(transport.calls).toHaveLength(1);
  });

  it("retries a 429 using the reset timestamp, then succeeds", async () => {
    const transport = stubTransport(
      jsonResponse(429, {
        error: "Too many requests.",
        code: "RATE_LIMIT_IP",
        reset: Date.now() + 5000,
      }),
      jsonResponse(200, { prompts: [] }),
    );
    const { client, delays } = makeClient(transport);

    await expect(client.listPrompts()).resolves.toEqual({ prompts: [] });

    expect(transport.calls).toHaveLength(2);
    expect(delays).toHaveLength(1);
    expect(delays[0]).toBeGreaterThanOrEqual(4000);
    expect(delays[0]).toBeLessThanOrEqual(5000);
  });

  it("retries transient 5xx failures with exponential backoff", async () => {
    const transport = stubTransport(
      jsonResponse(503, { error: "degraded" }),
      jsonResponse(503, { error: "degraded" }),
      jsonResponse(200, { ok: true }),
    );
    const { client, delays } = makeClient(transport, {
      retryBaseDelayMs: 100,
      retryMaxDelayMs: 1000,
    });

    await expect(client.get("/api/prompts")).resolves.toEqual({ ok: true });

    expect(transport.calls).toHaveLength(3);
    expect(delays).toHaveLength(2);
    expect(delays[0]).toBeGreaterThanOrEqual(100);
    expect(delays[1]).toBeGreaterThanOrEqual(200);
  });

  it("retries an in-flight idempotency lock", async () => {
    const transport = stubTransport(
      jsonResponse(409, {
        error: "A request with this Idempotency-Key is still being processed.",
      }),
      jsonResponse(200, { message: "ok" }),
    );
    const { client } = makeClient(transport);

    await expect(
      client.post("/api/prompts", {
        body: { title: "t" },
        idempotencyKey: "k-1",
        retry: 1,
      }),
    ).resolves.toEqual({ message: "ok" });

    expect(transport.calls).toHaveLength(2);
    expect(transport.calls[1].init.headers["Idempotency-Key"]).toBe("k-1");
  });

  it("honours `retry: false`", async () => {
    const transport = stubTransport(jsonResponse(503, { error: "degraded" }));
    const { client, delays } = makeClient(transport);

    await expect(
      client.get("/api/prompts", { retry: false }),
    ).rejects.toBeInstanceOf(PromptHashApiError);

    expect(transport.calls).toHaveLength(1);
    expect(delays).toHaveLength(0);
  });

  it("stops after the configured retry budget", async () => {
    const transport = stubTransport(jsonResponse(500, { error: "boom" }));
    const { client, delays } = makeClient(transport, { maxRetries: 3 });

    await expect(client.get("/api/prompts")).rejects.toBeInstanceOf(
      PromptHashApiError,
    );

    expect(transport.calls).toHaveLength(4);
    expect(delays).toHaveLength(3);
  });

  it("wraps transport failures in a network error", async () => {
    const transport = stubTransport(new Error("ECONNRESET"));
    const { client, delays } = makeClient(transport);

    const failure = client.get("/api/prompts");

    await expect(failure).rejects.toBeInstanceOf(PromptHashNetworkError);
    await failure.catch((err: PromptHashNetworkError) => {
      expect(err.message).toBe("ECONNRESET");
      expect(err.method).toBe("GET");
      expect(err.url).toBe("https://api.example.test/api/prompts");
    });
    expect(transport.calls).toHaveLength(3);
    expect(delays).toHaveLength(2);
  });
});

describe("PromptHashServerClient resource helpers", () => {
  it("registers a webhook subscription", async () => {
    const transport = stubTransport(
      jsonResponse(201, { id: "w1", secret: "s3cret" }),
    );
    const { client } = makeClient(transport);

    const result = await client.registerWebhook({
      walletAddress: "GABC",
      url: "https://example.test/hook",
      events: ["PromptPurchased"],
    });

    expect(result.secret).toBe("s3cret");
    expect(transport.calls[0].url).toBe("https://api.example.test/api/webhooks");
    expect(JSON.parse(transport.calls[0].init.body ?? "{}")).toEqual({
      walletAddress: "GABC",
      url: "https://example.test/hook",
      events: ["PromptPurchased"],
    });
  });

  it("scopes webhook reads by wallet", async () => {
    const transport = stubTransport(jsonResponse(200, { url: "https://x" }));
    const { client } = makeClient(transport);

    await client.listWebhookDeliveries("GABC");

    expect(transport.calls[0].url).toBe(
      "https://api.example.test/api/webhooks/deliveries?walletAddress=GABC",
    );
  });

  it("encodes path segments", async () => {
    const transport = stubTransport(jsonResponse(200, { id: "1" }));
    const { client } = makeClient(transport);

    await client.getPrompt("prompt 1/2");

    expect(transport.calls[0].url).toBe(
      "https://api.example.test/api/prompts/prompt%201%2F2",
    );
  });
});
